gensim/matutils.py

```diff
@@ -382,6 +383,112 @@ def cossim(vec1, vec2):
+            dense2 = sparse2full(vec2, max_len)
+            return scipy.stats.entropy(dense1, dense2)
+        else:
+            max_len = max(len(vec1), len(vec2))
+            dense1 = sparse2full(vec1, max_len)
+            dense2 = sparse2full(vec2, max_len)
+            return scipy.stats.entropy(dense1, dense2)
+    else:
+        if len(vec1) == 1:
+            vec1 = vec1[0]
+        if len(vec2) == 1:
+            vec2 = vec2[0]
+        # this conversion is made because if it is not in bow format, it might be a list within a list.
+        # the scipy implementation of Kullback fails in such a case so we pick up only the nested list.
+        return scipy.stats.entropy(vec1, vec2)
+
+
+def hellinger(vec1, vec2):
+    """
+    Hellinger distance is a distance metric to quantify the similarity between two probability distributions.
+    Similarity between distributions will be a number between <0,1>, where 0 is maximum similarity and 1 is minimum similarity.
+    """
+    if scipy.sparse.issparse(vec1):
+        vec1 = numpy.asarray(vec1.todense())
+    if scipy.sparse.issparse(vec2):
+        vec2 = numpy.asarray(vec2.todense())
+    if isbow(vec1) and isbow(vec2): 
+        # if it is a bag of words format, instead of converting to dense we use dictionaries to calculate appropriate distance
+        vec1, vec2 = dict(vec1), dict(vec2)
+        if len(vec2) < len(vec1): 
+            vec1, vec2 = vec2, vec1 # swap references so that we iterate over the shorter vector
+        sim = numpy.sqrt(0.5*sum((numpy.sqrt(value) - numpy.sqrt(vec2.get(index, 0.0)))**2 for index, value in iteritems(vec1)))
+        return sim
+    else:
+        sim = numpy.sqrt(0.5 * ((numpy.sqrt(vec1) - numpy.sqrt(vec2))**2).sum())
+        return sim
+
+def jaccard(vec1,vec2):
+    """
+    A similarity metric between bags of words representation.
+    Return the intersection divided by union, where union is the sum of the size of the two bags.
+    Highest value jaccard similarity of two bags is 1/2, which indicates the highest similarity.
+    If it is not a bag of words representation, the union and intersection is calculated in the traditional manner.
+    Returns a similarity in range <0,1> where values closer to 1 mean a higher similarity.
+    """
+    if scipy.sparse.issparse(vec1):
+        vec1 = numpy.asarray(vec1.todense())
+    if scipy.sparse.issparse(vec2):
+        vec2 = numpy.asarray(vec2.todense())
+    # converting from sparse for easier manipulation    
+    if isbow(vec1) and isbow(vec2): 
+        # if it's in bow format, we use the following definitions:
+        # union = sum of the 'weights' of both the bags
+        # intersection = lowest weight for a particular id; basically the number of common words or items 
+        intersection = 0
+        union = sum(weight for id_, weight in vec1) + sum(weight for id_, weight in vec2)
+        vec1, vec2 = dict(vec1), dict(vec2)
+        for item in vec1:
+            if item in vec2:
```

Comment:
This nested loop looks like a bug -- the outer `item` is replaced by the inner `item`??

The unit tests should definitely catch this.